### DISCUSSION TITLE
Why doesn't allow covariation here？

### DISCUSSION THREAD

**Discussion by LeaFrock** [dotnet/csharplang#3985] [General] [answered]
I find an interesting thing and I'm confused whether it's due to limit of compiler or is just by design. Codes see below,

```C#
    public class SingleHashFunctionList<T> : IHashFunctionList  
        where T : IHashFunction //Notice this 'where'
    {
        private readonly IReadOnlyList<T> _hashFunctions;

        public SingleHashFunctionList()
        {
            // Skip init work of '_hashFunctions'
            HashFunctions = _hashFunctions; //Compiler error: ...cannot implictly convert...
        }

        public IReadOnlyList<IHashFunction> HashFunctions { get; }
    }
```

Since the interface `IReadOnlyList<out T>` supports covariation and the compiler should know `T` must inherit the interface `IHashFunction`, why is `IReadOnlyList<T>` disallowed to be implictly converted to `IReadOnlyList<IHashFunction>` ?

**Comment by YairHalberstadt** [comment, marked as answer]
You need to add the class constraint, since covariant type parameters don't work with structs:

```csharp
    public class SingleHashFunctionList<T> : IHashFunctionList  
        where T : class, IHashFunction
    {
        private readonly IReadOnlyList<T> _hashFunctions;

        public SingleHashFunctionList()
        {
            // Skip init work of '_hashFunctions'
            HashFunctions = _hashFunctions;
        }

        public IReadOnlyList<IHashFunction> HashFunctions { get; }
    }
```

**Reply by LeaFrock** [reply]
Wow thanks! Maybe I need to create an issue for dotnet/roslyn that the analyzer should be smarter and the error message can be more friendly here...

**Reply by yaakov-h** [reply]
I, for one, would welcome an improved diagnostic.
